util/lsp/src/gotodefinition.rs

```diff
@@ -0,0 +1,167 @@
+//! Implementation of the goto-definition functionality of the LSP server
+
+use tower_lsp::{jsonrpc, lsp_types::*};
+
+use super::conversion::*;
+use super::server::*;
+use query::*;
+
+pub async fn goto_definition(
+    server: &Server,
+    params: GotoDefinitionParams,
+) -> jsonrpc::Result<Option<GotoDefinitionResponse>> {
+    let pos_params = params.text_document_position_params;
+    let text_document = pos_params.text_document;
+
+    server
+        .client
+        .log_message(MessageType::INFO, format!("GotoDefinition request: {}", text_document.uri))
+        .await;
+
+    let pos = pos_params.position;
+    let db = server.database.read().await;
+    let index = db.get(&text_document.uri).unwrap();
+    let info =
+        index.location_to_index(pos.from_lsp()).and_then(|idx| index.hoverinfo_at_index(idx));
+    let res = info.and_then(|info| info_to_jump(&index, info));
+    Ok(res)
+}
+
+fn info_to_jump(index: &DatabaseView, info: Info) -> Option<GotoDefinitionResponse> {
+    info.content.to_jump_content(index).map(GotoDefinitionResponse::Scalar)
+}
+
+trait ToJumpContent {
+    fn to_jump_content(&self, index: &DatabaseView) -> Option<Location>;
+}
+
+impl ToJumpContent for InfoContent {
+    fn to_jump_content(&self, index: &DatabaseView) -> Option<Location> {
+        match self {
+            InfoContent::VariableInfo(i) => i.to_jump_content(index),
+            InfoContent::TypeCtorInfo(i) => i.to_jump_content(index),
+            InfoContent::CallInfo(i) => i.to_jump_content(index),
+            InfoContent::DotCallInfo(i) => i.to_jump_content(index),
+            InfoContent::TypeUnivInfo(i) => i.to_jump_content(index),
+            InfoContent::HoleInfo(i) => i.to_jump_content(index),
+            InfoContent::AnnoInfo(i) => i.to_jump_content(index),
+            InfoContent::LocalComatchInfo(i) => i.to_jump_content(index),
+            InfoContent::LocalMatchInfo(i) => i.to_jump_content(index),
+            InfoContent::DefInfo(i) => i.to_jump_content(index),
+            InfoContent::CodefInfo(i) => i.to_jump_content(index),
+            InfoContent::LetInfo(i) => i.to_jump_content(index),
+            InfoContent::DataInfo(i) => i.to_jump_content(index),
+            InfoContent::CodataInfo(i) => i.to_jump_content(index),
+            InfoContent::CtorInfo(i) => i.to_jump_content(index),
+            InfoContent::DtorInfo(i) => i.to_jump_content(index),
+        }
+    }
+}
```

Comment:
This code could be simplified by replacing all cases that evaluate to `None` with the wildcard pattern `_ => None`. I would also remove the trait and implement `to_jump_content` as a method on `InfoContent` or as free-standing function.

Reply:
I have simplified the exhaustive pattern match by replacing it with a wildcard pattern. I have left the trait for the moment, but simplified it a bit.